segmentstore/server/src/main/java/io/pravega/segmentstore/server/tables/KeyTranslator.java

```diff
@@ -0,0 +1,207 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.server.tables;
+
+import com.google.common.base.Preconditions;
+import io.pravega.common.util.ArrayView;
+import io.pravega.common.util.ByteArraySegment;
+import io.pravega.segmentstore.contracts.tables.TableEntry;
+import io.pravega.segmentstore.contracts.tables.TableKey;
+import lombok.RequiredArgsConstructor;
+
+/**
+ * Translates Table Segment Keys from an external form into an internal one and back.
+ */
+abstract class KeyTranslator {
```

Comment:
What's this `KeyTranslator` about?

Reply:
I need to separate user keys from the Sorted Key Index keys (since I store the Sorted Key Index alongside the rest of the keys). I do not want internal and external keys to clash, or for a user to (accidentally) modify a SKI key-value or viceversa.

For Sorted Table Segments only, this translator adds a 1-byte prefix to each key, as follows:
- User-keys: byte equivalent of `E` (for External)
- SKI keys: byte equivalent of `I` (for Internal).

For Hash Table Segments, the no-op translator is used which doesn't do anything (so no effect on existing keys). 

Currently this prepend is inefficient due to a mandatory array copy. However with #4837 this will be a simple, zero-copy composition which has no overhead.